File: core/SumOfSubarrayRanges/SumOfSubarrayRanges.java
package com.dbc;

import java.util.Stack;

public class SumOfSubarrayRanges {
    public long subArrayRanges(int[] nums) {
        int[] minLeft = new int[nums.length];
        int[] maxLeft = new int[nums.length];
        int[] minRight = new int[nums.length];
        int[] maxRight = new int[nums.length];
        Stack<Integer> stack1 = new Stack<>();
        Stack<Integer> stack2 = new Stack<>();

        for (int i = 0; i < nums.length; i++) {
            while (!stack1.isEmpty() && nums[stack1.peek()] > nums[i]) {
                stack1.pop();
            }
            minLeft[i] = stack1.isEmpty() ? -1 : stack1.peek();
            stack1.push(i);

            while (!stack2.isEmpty() && nums[stack2.peek()] < nums[i]) {
                stack2.pop();
            }
            maxLeft[i] = stack2.isEmpty() ? -1 : stack2.peek();
            stack2.push(i);
        }

        stack1 = new Stack<>();
        stack2 = new Stack<>();
        for (int i = nums.length - 1; i >= 0; i--) {
            while (!stack1.isEmpty() && nums[stack1.peek()] >= nums[i]) {
                stack1.pop();
            }
            minRight[i] = stack1.isEmpty() ? nums.length : stack1.peek();
            stack1.push(i);

            while (!stack2.isEmpty() && nums[stack2.peek()] <= nums[i]) {
                stack2.pop();
            }
            maxRight[i] = stack2.isEmpty() ? nums.length : stack2.peek();
            stack2.push(i);
        }

        long minSum = 0, maxSum = 0;
        for (int i = 0; i < nums.length; i++) {
            minSum += (long) (i - minLeft[i]) * (minRight[i] - i) * nums[i];
            maxSum += (long) (i - maxLeft[i]) * (maxRight[i] - i) * nums[i];
        }

        return maxSum - minSum;
    }
}

File: core/SumOfSubarrayRanges/SumOfSubarrayRanges.py
from typing import List

def subArrayRanges(nums: List[int]) -> int:
    min_left, max_left, min_right, max_right = [], [], [], []
    stack1, stack2 = [], []
    for i in range(len(nums)):
        while stack1 and nums[stack1[-1]] >= nums[i]:
            stack1.pop()

        min_left.append(stack1[-1] if stack1 else -1)
        stack1.append(i)

        while stack2 and nums[stack2[-1]] <= nums[i]:
            stack2.pop()

        max_left.append(stack2[-1] if stack2 else -1)
        stack2.append(i)

    stack1, stack2 = [], []
    for i in range(len(nums) - 1, -1, -1):
        while stack1 and nums[stack1[-1]] > nums[i]:
            stack1.pop()

        min_right.insert(0, stack1[-1] if stack1 else len(nums))
        stack1.append(i)

        while stack2 and nums[stack2[-1]] < nums[i]:
            stack2.pop()

        max_right.insert(0, stack2[-1] if stack2 else len(nums))
        stack2.append(i)

    min_sum = sum([(right - index) * (index - left) * nums[index] for index, (left, right) in enumerate(zip(min_left, min_right))])
    max_sum = sum([(right - index) * (index - left) * nums[index] for index, (left, right) in enumerate(zip(max_left, max_right))])

    return max_sum - min_sum

if __name__ == '__main__':
    print(subArrayRanges([1,3,3]))
